common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -66,6 +66,13 @@ public static class Collation {
      */
     public final boolean isBinaryCollation;
 
+    /**
+     * Built-in string functions with collation support.
```

Comment:
Instead of adding new functions here I would propose following:
1) Add only low level basic functions here (e.g. matchAt).
2) Extend UTF8String with   `public boolean startsWith(final UTF8String prefix, *collationId*)`.  Have all the routing happen there.
3) Short circuit if we have `DEFAULT_COLLATION_ID` to previous implementation (e.g. (startsWith(final UTF8String prefix)). That way we will not risk regressing previous version.
4) Have operators call startsWith/endsWith... with proper collationId.

Reply:
Sounds good, on it!

Reply:
Update: as agreed in another comment thread, routing is now partially moved over to UTF8String, while binary collations are resolved right away in Scala in order to prevent performance regression